### PR TITLE
Make sure only one package syncing operation happens at a time

### PR DESCRIPTION
**Pull request by kpratyus**
This is the revised PR for issue https://github.com/open-telemetry/opamp-go/issues/84 . As proposed I am in the process to write a test case for testing the changes (It will take some time as I am new to the codebase).

### CODE REVIEW

**Comment by tigrannajaryan** [comment]
> Hi @tigrannajaryan , one case which could lead to race is when a package having same name, version and hash as previous package is tried to be Installed. Ideally the packagesyncer should ignore the request but due to race condition it could try to install it again. Do you have any other cases on top of your mind?

I think the most common case we want to test for is this:
1. Agent receives a ServerToAgent message with PackagesAvailable field set.
2. Agent creates a corresponding PackagesSyncer and begins downloading and syncing the packages. This happen in packagesSyncer.doSync on a different goroutine.
3. Agent receives another ServerToAgent message with PackagesAvailable field set. The value of PackagesAvailable may be different than the first time (i.e. something has changed on the Server side, while step 2 was executing.
4. Agent creates another PackagesSyncer and calls Sync(). This syncing should wait until the previous one in Step 2 is complete.

File: client/internal/clientcommon.go [hunk 1]
```diff
@@ -54,6 +54,9 @@ type ClientCommon struct {
 
 	// Indicates that the Client is fully stopped.
 	stoppedSignal chan struct{}
+
+	// Mutex for Package Syncing
+	PackageSyncComplete sync.Mutex
```
Comment (tigrannajaryan) [inline client/internal/clientcommon.go:59]: Nit: we use "Mutex" prefix elsewhere in this codebase.
```suggestion
	PackageSyncMutex sync.Mutex
```

**Review by tigrannajaryan — COMMENTED** [review]
Looks good so far. Let's add a test and should be good to merge

**Comment by kpratyus** [comment]
Facing the following race condition. Trying to debug it. Any help would be appreciated.

```
Read at 0x00c000380da8 by goroutine 24:
  runtime.slicecopy()
      /usr/lib/golang/src/runtime/slice.go:284 +0x0
  google.golang.org/protobuf/internal/impl.mergeBytesNoZero()
      /root/go/pkg/mod/google.golang.org/protobuf@v1.27.1/internal/impl/merge.go:167 +0xc6
  google.golang.org/protobuf/internal/impl.(*MessageInfo).mergePointer()
      /root/go/pkg/mod/google.golang.org/protobuf@v1.27.1/internal/impl/merge.go:52 +0x2b5
  google.golang.org/protobuf/internal/impl.(*MessageInfo).merge()
      /root/go/pkg/mod/google.golang.org/protobuf@v1.27.1/internal/impl/merge.go:32 +0x9a
  google.golang.org/protobuf/internal/impl.(*MessageInfo).merge-fm()
      /root/go/pkg/mod/google.golang.org/protobuf@v1.27.1/internal/impl/merge.go:23 +0x84
  google.golang.org/protobuf/proto.mergeOptions.mergeMessage()
      /root/go/pkg/mod/google.golang.org/protobuf@v1.27.1/proto/merge.go:73 +0xc8
  google.golang.org/protobuf/proto.Clone()
      /root/go/pkg/mod/google.golang.org/protobuf@v1.27.1/proto/merge.go:58 +0x84
  github.com/open-telemetry/opamp-go/client/internal.(*ClientSyncedState).SetPackageStatuses()
      /opamp-go/client/internal/clientstate.go:121 +0x67
  github.com/open-telemetry/opamp-go/client/internal.(*packagesSyncer).reportStatuses()
      /opamp-go/client/internal/packagessyncer.go:336 +0x1c6
  github.com/open-telemetry/opamp-go/client/internal.(*packagesSyncer).syncPackage()
      /opamp-go/client/internal/packagessyncer.go:223 +0xa5e
  github.com/open-telemetry/opamp-go/client/internal.(*packagesSyncer).doSync()
      /opamp-go/client/internal/packagessyncer.go:127 +0x4e5
  github.com/open-telemetry/opamp-go/client/internal.(*packagesSyncer).Sync·dwrap·11()
      /opamp-go/client/internal/packagessyncer.go:65 +0x64

Previous write at 0x00c000380da8 by goroutine 72:
  runtime.slicecopy()
      /usr/lib/golang/src/runtime/slice.go:284 +0x0
  google.golang.org/protobuf/internal/impl.consumeBytesNoZero()
      /root/go/pkg/mod/google.golang.org/protobuf@v1.27.1/internal/impl/codec_gen.go:5410 +0xe5
  google.golang.org/protobuf/internal/impl.(*MessageInfo).unmarshalPointer()
      /root/go/pkg/mod/google.golang.org/protobuf@v1.27.1/internal/impl/decode.go:139 +0x79d
  google.golang.org/protobuf/internal/impl.consumeMessageInfo()
      /root/go/pkg/mod/google.golang.org/protobuf@v1.27.1/internal/impl/codec_field.go:252 +0x25c
  google.golang.org/protobuf/internal/impl.(*MessageInfo).unmarshalPointer()
      /root/go/pkg/mod/google.golang.org/protobuf@v1.27.1/internal/impl/decode.go:139 +0x79d
  google.golang.org/protobuf/internal/impl.(*MessageInfo).unmarshal()
      /root/go/pkg/mod/google.golang.org/protobuf@v1.27.1/internal/impl/decode.go:62 +0x104
  google.golang.org/protobuf/internal/impl.(*MessageInfo).unmarshal-fm()
      /root/go/pkg/mod/google.golang.org/protobuf@v1.27.1/internal/impl/decode.go:55 +0x104
  google.golang.org/protobuf/proto.UnmarshalOptions.unmarshal()
      /root/go/pkg/mod/google.golang.org/protobuf@v1.27.1/proto/decode.go:93 +0x2bd
  google.golang.org/protobuf/proto.Unmarshal()
      /root/go/pkg/mod/google.golang.org/protobuf@v1.27.1/proto/decode.go:50 +0xa4
  github.com/open-telemetry/opamp-go/client/internal.(*wsReceiver).receiveMessage()
      /opamp-go/client/internal/wsreceiver.go:73 +0x67
  github.com/open-telemetry/opamp-go/client/internal.(*wsReceiver).ReceiverLoop()
      /opamp-go/client/internal/wsreceiver.go:55 +0xf7
  github.com/open-telemetry/opamp-go/client.(*wsClient).runOneCycle()
      /opamp-go/client/wsclient.go:242 +0x5cf
  github.com/open-telemetry/opamp-go/client.(*wsClient).runUntilStopped()
      /opamp-go/client/wsclient.go:264 +0x44
  github.com/open-telemetry/opamp-go/client.(*wsClient).runUntilStopped-fm()
      /opamp-go/client/wsclient.go:257 +0x4d
  github.com/open-telemetry/opamp-go/client/internal.(*ClientCommon).StartConnectAndRun.func1()
      /opamp-go/client/internal/clientcommon.go:199 +0xb2
```

**Comment by srikanthccv** [comment]
@kpratyus are you still interested in continuing this work?

**Comment by tigrannajaryan** [comment]
Closing since there hasn't been updates to the PR for a while. Please reopen if still active.

**Pull request closed without merging**